dev/TEAL.md

```diff
@@ -0,0 +1,219 @@
+---
+numbersections: true
+title: "Algorand Transaction Execution Approval Language"
+date: \today
+abstract: >
+  Algorand allows transactions to be effectively signed by a small program. If the program evaluates to true then the transaction is allowed. This document defines the language and bytcode format.
+---
+
+# Transaction Execution Approval Language (TEAL)
+
+TEAL is a bytecode based stack language that executes inside Algorand transactions to check the parameters of the transaction and approve the transaction as if by a signature.
+
+TEAL programs should be short, at most 1000 bytes including all constants and operations, and run fast as they are run in-line along with signature checking, transaction balance rule checking, and other checks during block assembly and validation.
+
+## The Stack
+
+The stack starts empty and contains values of either uint64 or bytes. (`bytes` implemented in Go as a []byte slice) Most operations act on the stack, popping arguments from it and pushing results to it.
+
+The maximum stack depth is currently 1000.
+
+## Scratch Space
+
+In addition to the stack there are 256 positions of scratch space, also uint64-bytes union values, accessed by the `load` and `store` ops moving data from or to scratch space, respectively.
+
+## Execution Environment
+
+TEAL runs in Algorand nodes as part of testing a proposed transaction to see if it is valid and authorized to be committed into a block. If an authorized program executes and finishes with a single non-zero uint64 value on the stack then that program has validated the transaction it is attached to.
+
+A program is an authorized program one of two ways: The SHA512_256 hash of the program (prefixed by "Program") is equal to the transaction Sender address; OR the program (prefixed by "Program") has a valid signature or multi-signature from the transaction Sender address.
+
+The TEAL bytecode plus the length of any Args must add up to less than 1000 bytes. Each TEAL op has an associated cost estimate and the program cost estimate must total less than 20000. Most ops have an estimated cost of 1, but a few slow crypto ops are much higher.
```

Comment:
May also want to parametrize 2000 (e.g., `MaxProgramCost`)

Reply:
noted consensus param names where appropriate